config/configmapprovider/composite.go

```diff
@@ -0,0 +1,294 @@
+		watchContext:       ctx,
+		cancelWatchContext: cancelFunc,
+	}
+
+	return m, nil
+}
+
+func (cr *compositeRetrieved) Get() *config.Map {
+	// WatchableRetrieved requires that Get() is never called concurrently
+	// with WatchForUpdate() which means a race on cr.composedMap is not possible
+	// since the only place where we modify it is in WatchForUpdate(), so there
+	// is no need to protect cr.composedMap from races.
+	return cr.composedMap
+}
+
+func (cr *compositeRetrieved) WatchForUpdate() error {
+	// Create one go routine per sub-retrieved and start waiting for WatchForUpdate()
+	// of any Watchable sub-retrieved to return.
+
+	// Channel used to signal that sub-retrieved WaitForUpdate() returned.
+	readyWatchableCh := make(chan readyWatchable)
+
+	// Count how many of the sub-retrieved are watchable.
+	watchableCount := 0
+	for _, subRetrieved := range cr.subRetrieved {
+		if watchable, ok := subRetrieved.(WatchableRetrieved); ok {
+			// This is a Watchable sub-retrieved.
+			// Start watching for updates from this WatchableRetrieved.
+			watchableCount++
+			go func() {
+				err := watchable.WatchForUpdate()
+				select {
+				case readyWatchableCh <- readyWatchable{watchable, err}:
+					// We have an update, signal it to the outer func.
+				case <-cr.watchContext.Done():
+					// We are instructed to terminate. Just return.
+				}
+			}()
+		}
+	}
+
+	select {
+	case readyWatchable := <-readyWatchableCh:
+
+		// At least one watcher fired. Close all sub-Retrieved. We don't need
+		// to close the one that fired the signal (readyWatchable.watchable),
+		// so provide it as an exception to closeAllExcept().
+		if err := cr.closeAllExcept(cr.watchContext, readyWatchable.watchable); err != nil {
+			return err
+		}
+
+		// Wait for all sub-retrieved WatchForUpdate() to return. They either will return
+		// because there is an update or because we closed them via closeAllExcept()
+		// call above.
+		err := readyWatchable.returnedError
+		for i := 0; i < watchableCount-1; i++ {
+			select {
+			case watchable := <-readyWatchableCh:
+				// One more sub-retrieved's WatchForUpdate() returned.
+				// ErrSessionClosed is expected because we called Close. If it is
+				// any other error then it is unexpected so return the error to the caller.
+				if watchable.returnedError != nil && watchable.returnedError != configsource.ErrSessionClosed {
+					err = multierr.Append(err, watchable.returnedError)
+				}
+				continue
+			case <-cr.watchContext.Done():
+				// We are instructed to terminate.
+				return configsource.ErrSessionClosed
+			}
+		}
+
+		if err == nil {
+			// All Watchables returned without error, which means they are ready for
+			// another Get() call. Call Get() for all subproviders again and combine
+			// the results.
+			configMap, err2 := cr.subRetrieved.getAllAndCompose(cr.composeFunc)
+			if err2 != nil {
+				return err2
+			}
+
+			// Remember the result.
+			cr.composedMap = configMap
+		}
```

Comment:
I think this is an overoptimization. You just signal the caller that you got an update and let the collector redo the map etc. This will simplify the requirements of concurrent calls, and also the logic here.